specs/commandline/blob.md

```diff
@@ -0,0 +1,362 @@
+
+# Add a default signing key referencing the remote key identifier, and the plugin associated with it.
+notation key add --default --name <key_name> --plugin <plugin_name> --id <remote_key_id>
+
+# sign a blob
+notation blob sign --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+An example for a successful signing:
+
+```console
+$ notation blob sign --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+Successfully signed /tmp/my-blob.bin
+Signature written to /tmp/my-blob-signature.sig
+```
+
+### Sign a blob with on-demand remote key
+
+```shell
+notation blob sign --plugin <plugin_name> --id <remote_key_id> --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+### Sign a blob using COSE signature format
+
+```shell
+# Prerequisites:
+# A default signing key is configured using CLI "notation key"
+
+# Use option "--signature-format" to set the signature format to COSE.
+notation blob sign --signature-format cose --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+### Sign a blob using the default signing key
+
+```shell
+# Prerequisites:
+# A default signing key is configured using CLI "notation key"
+
+notation blob sign --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+### Sign a blob with user metadata
+
+```shell
+# Prerequisites:
+# A default signing key is configured using CLI "notation key"
+
+# sign a blob and add user-metadata io.wabbit-networks.buildId=123 to the payload
+notation blob sign --user-metadata io.wabbit-networks.buildId=123 --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+
+# sign a blob and add user-metadata io.wabbit-networks.buildId=123 and io.wabbit-networks.buildTime=1672944615 to the payload
+notation blob sign --user-metadata io.wabbit-networks.buildId=123 --user-metadata io.wabbit-networks.buildTime=1672944615 --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+### Sign a blob with media type
+
+```shell
+notation blob sign --media-type <media type> --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+### Sign a blob and specify the signature expiry duration, for example 24 hours
+
+```shell
+notation blob sign --expiry 24h --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+### Sign a blob using a specified signing key
+
+```shell
+# List signing keys to get the key name
+notation key list
+
+# Sign a container image using the specified key name
+notation blob sign --key <key_name> --signature /tmp/my-blob-signature.sig /tmp/my-blob.bin
+```
+
+## Inspect detached blob signatures
+
+### Display details of the given detached blob signature and its associated certificate properties
+
+
+```text
+notation blob inspect [flags] /tmp/my-blob-signature.sig
+```
+
+### Inspect the given detached blob signature
+
+```shell
+# Prerequisites: Signatures is produced by notation blob sign command
+notation blob inspect /tmp/my-blob-signature.sig
+```
+
+An example output:
+```shell
+Inspecting /tmp/my-blob-signature.sig
```

Comment:
How does `notation` detect the signature format? Should `notation` print out the signature format (`jws` / `cose`)?

Reply:
Same question for `verify`.

Reply:
Good question. I suggest Notation gets the signature format from the file extension. Something like `my-blob-signature.sig.cose`, `my-blob-signature.sig.jws`

For OCI, the format is an unsigned attribute that is present in the signature manifest (in other words, outside the signature envelope). Similarly, for Blobs, the format will come from outside the envelope i.e. file extension.

An alternative is encoding the format within the signature envelope file. Either json encoding or some sort of armoring like below.

```
{
    "mediaType": "application/cose",
    "signatureEnvelope": "<Base64 encoded envelope>"
}
``` 

or

```
-----START COSE SIGNATURE----
<Base64 encoded envelope>
-----END COSE SIGNATURE----
```

However, these two options will require Notation to parse the signature file in order to read the envelope. And also, tools other than Notation will not work out of the box as they do not understand the wrapping method,

Reply:
@rgnote Quick question on the file extension method:
Say a verifier receives a signature with name `my-blob-signature.sig`, how would notation tell which signature envelope format should be used? Because a signer can always rename their signature files after signing. 
Basically, my concern here is that Notation verification is going to be dependent on the signature's file name, which is not strictly controlled by Notation.
/cc: @priteshbandi

Reply:
@rgnote I have a question about the UX. 

```sh
SIGNATURE_NAME=blob.sig
notation blob sign --signature blob.sig ./filename
```
Then, when the user wants to verify the signature:
```sh
notation blob verify $SIGNATURE_NAME.jws
```
How does the user determine the suffix of the signature? I believe this design elevates the complexity of using the Notation CLI. Users will either have to use `ls` to scan the directory to find the actual name or delve into our documentation to understand the naming suffix.

Reply:
@Two-Hearts 

We discussed about scenario on the call. Yes, a user can change the file name (including its extension) however they like. We can't assume that users won't change extensions. Since extension is only a hint about the format, this is not a big problem for Notation. The proposed solution is
1. If there is a valid extension, Notation will use it
2. If there is no valid extension, Notation will guess the format by bruteforcing all the available formats it knows (this is already being done in notation-core-go)
3. If none of the above methods work, Notation errors out saying unknown format.

Reply:
@JeyJeyGao 

We picked `--signature` flag when we had no idea about how to pass signature format data with the signature itself. Now that we decided to go with file extensions, we need to update the flag name. I suggest we change it to `--signature-name`.

For example, let's say user gives `xyz` for the `--signature-name` field. 
On the signing side, Notation will write the signature to `xyz.sig.cose` or `xyz.sig.jws` to the CWD. 

On the verification side, notation will expect `--signature-path` argument, to which a user need to provide the location of the signature. Like `notation blob verify ./xyz.sig.cose`.

Reply:
> 2. If there is no valid extension, Notation will guess the format by bruteforcing all the available formats it knows (this is already being done in notation-core-go)

I see, so we are keeping step 2. I thought we were going to remove it on the call. SGTM.

Reply:
@rgnote Then the UX will be:
```sh
SIGNATURE_NAME=xyz
notation blob sign --signature-name $SIGNATURE_NAME ./filename
```
Verify the signature:
```sh
notation blob verify $SIGNATURE_NAME.sig.jws
```
In other words, the user must know the complete name of the signature; otherwise, the notation blob verify command cannot locate the signature file. How can the user determine the full name of the signature? They can either obtain it from the output of the `notation blob sign command` or figure it out on their own.

@yizha1 @FeynmanZhou Could you provide some insights from the perspective of a PM?

Reply:
From a verifier point of view, I don't see big problems, since verifier needs to put the full path of the signature file. And the sign command will output the full file path. But it's better to describe it in the document, that the signature name will be added with suffixes based on the signature format that users choose.